File: aoc/2015/day01/fortran/solution.f90
! Advent of Code 2015 Day 01 - Fortran Solution Stub
program day01
    implicit none
    integer :: ios, lu, i, n
    logical :: first
    character(len=16384) :: line

    open(newunit=lu, file="../input.txt", status="old", iostat=ios)
    if (ios /= 0) then
        print *, "Input file not found"
        stop
    end if

    do
        read(lu, '(a)', iostat=ios) line
        if (ios /= 0) exit
    end do
    close(lu)

    n = 0
    first = .false.
    do i = 1, len_trim(line)
        if (line(i:i) == '(') then
            n = n + 1
        else if (line(i:i) == ')') then
            n = n - 1
        end if
        if (.not.first) then
            if (n == -1) then
                print *, 'Santa enters basement at', i
                first = .true.
            end if
        end if
    end do

    print *, 'Santa goes to floor', n

end program day01

File: aoc/2015/day01/python/solution.py
# Advent of Code 2015 Day 01 - Python Solution Stub


def main():
    try:
        with open("../input.txt", "r") as f:
            data = f.read()
        print("Santa goes to floor", data.count("(") - data.count(")"))
        print("Santa enters basement at", pos_extra_rparen(data))
    except FileNotFoundError:
        print("Input file not found")


def pos_extra_rparen(data):
    n = 0
    for i, c in enumerate(data):
        if c == "(":
            n += 1
        elif c == ")":
            n -= 1
        else:
            raise Exception("unexpected char")
        if n == -1:
            return i + 1
    return 0


if __name__ == "__main__":
    main()
